h2o-automl/src/main/java/ai/h2o/automl/AutoML.java

```diff
@@ -164,76 +166,142 @@ public AutoML() {
-    if (buildSpec.build_control.stopping_criteria.stopping_tolerance() == AUTO_STOPPING_TOLERANCE) {
-      buildSpec.build_control.stopping_criteria.set_default_stopping_tolerance_for_frame(_trainingFrame);
-      eventLog().info(Stage.Workflow, "Setting stopping tolerance adaptively based on the training frame: " +
-              buildSpec.build_control.stopping_criteria.stopping_tolerance());
-    } else {
-      eventLog().info(Stage.Workflow, "Stopping tolerance set by the user: " + buildSpec.build_control.stopping_criteria.stopping_tolerance());
-      double default_tolerance = AutoMLBuildSpec.AutoMLStoppingCriteria.default_stopping_tolerance_for_frame(_trainingFrame);
-      if (buildSpec.build_control.stopping_criteria.stopping_tolerance() < 0.7 * default_tolerance){
-        eventLog().warn(Stage.Workflow, "Stopping tolerance set by the user is < 70% of the recommended default of " + default_tolerance + ", so models may take a long time to converge or may not converge at all.");
+  private void validateInput(AutoMLInput input) {
+    if (null == DKV.getGet(input.training_frame))
+      throw new H2OIllegalArgumentException("No training data has been specified, either as a path or a key, or it is not available anymore.");
+
+    final Frame trainingFrame = DKV.getGet(input.training_frame);
+    final Frame validationFrame = DKV.getGet(input.validation_frame);
+    final Frame blendingFrame = DKV.getGet(input.blending_frame);
+    final Frame leaderboardFrame = DKV.getGet(input.leaderboard_frame);
+
+    Map<String, Frame> compatibleFrames = new LinkedHashMap<String, Frame>(){{
+      put("training", trainingFrame);
+      put("validation", validationFrame);
+      put("blending", blendingFrame);
+      put("leaderboard", leaderboardFrame);
+    }};
+    for (Map.Entry<String, Frame> entry : compatibleFrames.entrySet()) {
+      Frame frame = entry.getValue();
+      if (frame != null && frame.find(input.response_column) < 0) {
+        throw new H2OIllegalArgumentException("Response column '"+input.response_column+"' is not in the "+entry.getKey()+" frame.");
       }
     }
+
+    if (input.fold_column != null && trainingFrame.find(input.fold_column) < 0) {
+      throw new H2OIllegalArgumentException("Fold column '"+input.fold_column+"' is not in the training frame.");
+    }
+    if (input.weights_column != null && trainingFrame.find(input.weights_column) < 0) {
+      throw new H2OIllegalArgumentException("Weights column '"+input.weights_column+"' is not in the training frame.");
+    }
+
+    if (input.ignored_columns != null) {
+      List<String> ignoredColumns = new ArrayList<>(Arrays.asList(input.ignored_columns));
+      Map<String, String> doNotIgnore = new LinkedHashMap<String, String>(){{
+        put("response_column", input.response_column);
+        put("fold_column", input.fold_column);
+        put("weights_column", input.weights_column);
+      }};
+      for (Map.Entry<String, String> entry: doNotIgnore.entrySet()) {
```

Comment:
@seb-h2o when I see `for` with just `if` inside I want to use streams with filters:
```
doNotIgnore.entrySet().stream()
              .filter(entry -> entry.getValue() != null && ignoredColumns.contains(entry.getValue()))
              .forEach(entry -> {
                eventLog().info(Stage.DataImport, "Removing "+entry.getKey()+" '"+entry.getValue()+"' from list of ignored columns.");
                ignoredColumns.remove(entry.getValue())
              });
```

Reply:
yeah, difference is not that big tbh.
Also, I'm not a complete fan for `forEach` in streams. I like to know when there are side effects and where there are none: so I use streams with pure functions and `for` for side effects.

Reply:
@seb-h2o in functional programming `forEach` signifies side effects. `map` is for pure things.

Reply:
I know that. And that's why it doesn't bring much value in a language like Java.